File: server/src/arstream_server/protocol.py
"""arstream wire protocol v1 -- a faithful Python mirror of docs/PROTOCOL.md.

Speaks the exact same wire format as mobile/native/src/net/protocol.cpp.
Both were written independently against docs/PROTOCOL.md; the golden-byte
files under fixtures/ verify that the two haven't silently drifted apart
(see tests/test_protocol.py).

Unlike the C++ side, JSON-bodied messages (HELLO, VIDEO_CONFIG's json part,
STATUS, GOODBYE) are fully encoded/decoded here too -- stdlib json costs
nothing in Python, whereas on the C++ side it's deliberately left as opaque
bytes (see the note at the top of protocol.h).
"""

from __future__ import annotations

import json
import struct
from dataclasses import dataclass, field
from enum import IntEnum

PROTOCOL_VERSION_1 = 0x0001
HEADER_SIZE = 12

_HEADER_STRUCT = struct.Struct(">IBBHI")  # payload_length, msg_type, flags, protocol_version, sequence_number


class MsgType(IntEnum):
    HELLO = 0x01
    HELLO_ACK = 0x02
    CLOCK_SYNC_REQUEST = 0x03
    CLOCK_SYNC_RESPONSE = 0x04
    VIDEO_CONFIG = 0x05
    VIDEO_CHUNK = 0x10
    IMU_BATCH = 0x20
    POSE_SAMPLE = 0x30
    POINT_CLOUD = 0x31
    CAMERA_INTRINSICS = 0x32
    STATUS = 0x40
    GOODBYE = 0xF0


@dataclass
class Header:
    payload_length: int
    msg_type: int
    flags: int = 0
    protocol_version: int = PROTOCOL_VERSION_1
    sequence_number: int = 0


@dataclass
class ImuSample:
    sensor_type: int  # 0=accelerometer, 1=gyroscope
    timestamp_ns: int
    x: float
    y: float
    z: float


@dataclass
class Point:
    x: float
    y: float
    z: float
    confidence: float


@dataclass
class Message:
    header: Header
    payload: bytes


class ProtocolError(ValueError):
    pass


def encode_header(h: Header) -> bytes:
    return _HEADER_STRUCT.pack(h.payload_length, h.msg_type, h.flags, h.protocol_version, h.sequence_number)


def decode_header(data: bytes) -> Header:
    if len(data) < HEADER_SIZE:
        raise ProtocolError(f"header requires at least {HEADER_SIZE} bytes, got {len(data)}")
    payload_length, msg_type, flags, protocol_version, sequence_number = _HEADER_STRUCT.unpack_from(data, 0)
    return Header(payload_length, msg_type, flags, protocol_version, sequence_number)


def _wrap(msg_type: MsgType, seq: int, payload: bytes) -> bytes:
    header = Header(payload_length=len(payload), msg_type=int(msg_type), sequence_number=seq)
    return encode_header(header) + payload


def next_message(buffer: bytes) -> tuple[Message, int] | None:
    """Safely skips unknown msg_types too -- returns None if the buffer
    doesn't contain a complete message yet (the caller should wait for more bytes)."""
    if len(buffer) < HEADER_SIZE:
        return None
    header = decode_header(buffer)
    total = HEADER_SIZE + header.payload_length
    if len(buffer) < total:
        return None
    payload = buffer[HEADER_SIZE:total]
    return Message(header, payload), total


def iter_messages(buffer: bytes):
    """Yields every complete message in buffer, in order (generator)."""
    offset = 0
    while True:
        result = next_message(buffer[offset:])
        if result is None:
            return
        message, consumed = result
        yield message
        offset += consumed


# -- HELLO / HELLO_ACK / STATUS / GOODBYE (JSON body) --


def encode_hello(seq: int, body: dict) -> bytes:
    return _wrap(MsgType.HELLO, seq, json.dumps(body).encode("utf-8"))


def decode_hello(payload: bytes) -> dict:
    return json.loads(payload.decode("utf-8"))


def encode_hello_ack(seq: int, body: dict) -> bytes:
    return _wrap(MsgType.HELLO_ACK, seq, json.dumps(body).encode("utf-8"))


def decode_hello_ack(payload: bytes) -> dict:
    return json.loads(payload.decode("utf-8"))


def encode_status(seq: int, body: dict) -> bytes:
    return _wrap(MsgType.STATUS, seq, json.dumps(body).encode("utf-8"))


def decode_status(payload: bytes) -> dict:
    return json.loads(payload.decode("utf-8"))


def encode_goodbye(seq: int, body: dict) -> bytes:
    return _wrap(MsgType.GOODBYE, seq, json.dumps(body).encode("utf-8"))


def decode_goodbye(payload: bytes) -> dict:
    return json.loads(payload.decode("utf-8"))


# -- Clock synchronization --


def encode_clock_sync_request(seq: int, client_send_time_ns: int) -> bytes:
    return _wrap(MsgType.CLOCK_SYNC_REQUEST, seq, struct.pack(">q", client_send_time_ns))


def decode_clock_sync_request(payload: bytes) -> int:
    (client_send_time_ns,) = struct.unpack_from(">q", payload, 0)
    return client_send_time_ns


def encode_clock_sync_response(seq: int, client_send_time_ns: int, server_recv_time_ns: int, server_send_time_ns: int) -> bytes:
    return _wrap(
        MsgType.CLOCK_SYNC_RESPONSE,
        seq,
        struct.pack(">qqq", client_send_time_ns, server_recv_time_ns, server_send_time_ns),
    )


def decode_clock_sync_response(payload: bytes) -> tuple[int, int, int]:
    return struct.unpack_from(">qqq", payload, 0)


# -- Video --


def encode_video_config(seq: int, body: dict, sps_pps_annexb: bytes) -> bytes:
    json_bytes = json.dumps(body).encode("utf-8")
    payload = struct.pack(">H", len(json_bytes)) + json_bytes + sps_pps_annexb
    return _wrap(MsgType.VIDEO_CONFIG, seq, payload)


def decode_video_config(payload: bytes) -> tuple[dict, bytes]:
    (json_len,) = struct.unpack_from(">H", payload, 0)
    json_bytes = payload[2 : 2 + json_len]
    sps_pps = payload[2 + json_len :]
    return json.loads(json_bytes.decode("utf-8")), sps_pps


def encode_video_chunk(seq: int, capture_timestamp_ns: int, is_keyframe: bool, nal_data: bytes) -> bytes:
    payload = struct.pack(">qB", capture_timestamp_ns, 0x01 if is_keyframe else 0x00) + nal_data
    return _wrap(MsgType.VIDEO_CHUNK, seq, payload)


def decode_video_chunk(payload: bytes) -> tuple[int, bool, bytes]:
    capture_timestamp_ns, flags = struct.unpack_from(">qB", payload, 0)
    return capture_timestamp_ns, bool(flags & 0x01), payload[9:]


# -- IMU --


def encode_imu_batch(seq: int, samples: list[ImuSample]) -> bytes:
    payload = bytearray(struct.pack(">H", len(samples)))
    for s in samples:
        payload += struct.pack(">Bqfff", s.sensor_type, s.timestamp_ns, s.x, s.y, s.z)
    return _wrap(MsgType.IMU_BATCH, seq, bytes(payload))


def decode_imu_batch(payload: bytes) -> list[ImuSample]:
    (count,) = struct.unpack_from(">H", payload, 0)
    offset = 2
    samples = []
    for _ in range(count):
        sensor_type, timestamp_ns, x, y, z = struct.unpack_from(">Bqfff", payload, offset)
        samples.append(ImuSample(sensor_type, timestamp_ns, x, y, z))
        offset += 21
    return samples


# -- Optional ARCore fields --


def encode_pose_sample(seq: int, timestamp_ns: int, tracking_state: int, x: float, y: float, z: float, qx: float, qy: float, qz: float, qw: float) -> bytes:
    payload = struct.pack(">qBfffffff", timestamp_ns, tracking_state, x, y, z, qx, qy, qz, qw)
    return _wrap(MsgType.POSE_SAMPLE, seq, payload)


def decode_pose_sample(payload: bytes):
    return struct.unpack_from(">qBfffffff", payload, 0)


def encode_point_cloud(seq: int, timestamp_ns: int, points: list[Point]) -> bytes:
    payload = bytearray(struct.pack(">qI", timestamp_ns, len(points)))
    for p in points:
        payload += struct.pack(">ffff", p.x, p.y, p.z, p.confidence)
    return _wrap(MsgType.POINT_CLOUD, seq, bytes(payload))


def decode_point_cloud(payload: bytes) -> tuple[int, list[Point]]:
    timestamp_ns, count = struct.unpack_from(">qI", payload, 0)
    offset = 12
    points = []
    for _ in range(count):
        x, y, z, confidence = struct.unpack_from(">ffff", payload, offset)
        points.append(Point(x, y, z, confidence))
        offset += 16
    return timestamp_ns, points


def encode_camera_intrinsics(seq: int, fx: float, fy: float, cx: float, cy: float, width: int, height: int) -> bytes:
    payload = struct.pack(">ffffII", fx, fy, cx, cy, width, height)
    return _wrap(MsgType.CAMERA_INTRINSICS, seq, payload)


def decode_camera_intrinsics(payload: bytes):
    return struct.unpack_from(">ffffII", payload, 0)

File: mobile/native/src/net/protocol.cpp
#include "protocol.h"

#include <cstring>

namespace arstream::protocol {

namespace {

// Big-endian yazma/okuma yardimcilari -- OS soket basliklari (htonl vb.)
// KASITLI kullanilmiyor, boylece bu dosya host'ta soket kutuphanesi
// baglamadan da derlenir.

void put_u8(std::vector<uint8_t> &buf, uint8_t v) {
	buf.push_back(v);
}

void put_u16be(std::vector<uint8_t> &buf, uint16_t v) {
	buf.push_back(static_cast<uint8_t>((v >> 8) & 0xFF));
	buf.push_back(static_cast<uint8_t>(v & 0xFF));
}

void put_u32be(std::vector<uint8_t> &buf, uint32_t v) {
	buf.push_back(static_cast<uint8_t>((v >> 24) & 0xFF));
	buf.push_back(static_cast<uint8_t>((v >> 16) & 0xFF));
	buf.push_back(static_cast<uint8_t>((v >> 8) & 0xFF));
	buf.push_back(static_cast<uint8_t>(v & 0xFF));
}

void put_i64be(std::vector<uint8_t> &buf, int64_t v) {
	uint64_t u = static_cast<uint64_t>(v);
	for (int i = 7; i >= 0; --i) {
		buf.push_back(static_cast<uint8_t>((u >> (i * 8)) & 0xFF));
	}
}

void put_f32be(std::vector<uint8_t> &buf, float v) {
	uint32_t bits;
	memcpy(&bits, &v, sizeof(bits));
	put_u32be(buf, bits);
}

void put_bytes(std::vector<uint8_t> &buf, const uint8_t *data, size_t size) {
	buf.insert(buf.end(), data, data + size);
}

void put_bytes(std::vector<uint8_t> &buf, const std::string &s) {
	buf.insert(buf.end(), s.begin(), s.end());
}

bool get_u16be(const uint8_t *p, size_t size, size_t offset, uint16_t &out) {
	if (offset + 2 > size) {
		return false;
	}
	out = (static_cast<uint16_t>(p[offset]) << 8) | p[offset + 1];
	return true;
}

bool get_u32be(const uint8_t *p, size_t size, size_t offset, uint32_t &out) {
	if (offset + 4 > size) {
		return false;
	}
	out = (static_cast<uint32_t>(p[offset]) << 24) | (static_cast<uint32_t>(p[offset + 1]) << 16) | (static_cast<uint32_t>(p[offset + 2]) << 8) | p[offset + 3];
	return true;
}

bool get_i64be(const uint8_t *p, size_t size, size_t offset, int64_t &out) {
	if (offset + 8 > size) {
		return false;
	}
	uint64_t u = 0;
	for (int i = 0; i < 8; i++) {
		u = (u << 8) | p[offset + i];
	}
	out = static_cast<int64_t>(u);
	return true;
}

bool get_f32be(const uint8_t *p, size_t size, size_t offset, float &out) {
	uint32_t bits;
	if (!get_u32be(p, size, offset, bits)) {
		return false;
	}
	memcpy(&out, &bits, sizeof(out));
	return true;
}

// Header'i yazip ardindan payload'i ekleyen ortak govde.
std::vector<uint8_t> wrap(MsgType type, uint32_t seq, const std::vector<uint8_t> &payload) {
	Header h;
	h.payload_length = static_cast<uint32_t>(payload.size());
	h.msg_type = static_cast<uint8_t>(type);
	h.flags = 0;
	h.protocol_version = kProtocolVersion1;
	h.sequence_number = seq;

	std::vector<uint8_t> out = encode_header(h);
	out.insert(out.end(), payload.begin(), payload.end());
	return out;
}

} // namespace

std::vector<uint8_t> encode_header(const Header &h) {
	std::vector<uint8_t> buf;
	buf.reserve(kHeaderSize);
	put_u32be(buf, h.payload_length);
	put_u8(buf, h.msg_type);
	put_u8(buf, h.flags);
	put_u16be(buf, h.protocol_version);
	put_u32be(buf, h.sequence_number);
	return buf;
}

bool decode_header(const uint8_t *data, size_t size, Header &out) {
	if (size < kHeaderSize) {
		return false;
	}
	uint32_t payload_length, sequence_number;
	uint16_t protocol_version;
	if (!get_u32be(data, size, 0, payload_length)) {
		return false;
	}
	out.payload_length = payload_length;
	out.msg_type = data[4];
	out.flags = data[5];
	if (!get_u16be(data, size, 6, protocol_version)) {
		return false;
	}
	out.protocol_version = protocol_version;
	if (!get_u32be(data, size, 8, sequence_number)) {
		return false;
	}
	out.sequence_number = sequence_number;
	return true;
}

bool next_message(const uint8_t *buffer, size_t size, MessageView &out, size_t &bytes_consumed) {
	Header h;
	if (!decode_header(buffer, size, h)) {
		return false;
	}
	size_t total = kHeaderSize + h.payload_length;
	if (size < total) {
		return false; // henuz tam mesaj gelmemis
	}
	out.header = h;
	out.payload = buffer + kHeaderSize;
	out.payload_size = h.payload_length;
	bytes_consumed = total;
	return true;
}

std::vector<uint8_t> encode_hello(uint32_t seq, const std::string &json_body) {
	std::vector<uint8_t> payload;
	put_bytes(payload, json_body);
	return wrap(MsgType::Hello, seq, payload);
}

std::vector<uint8_t> encode_hello_ack(uint32_t seq, const std::string &json_body) {
	std::vector<uint8_t> payload;
	put_bytes(payload, json_body);
	return wrap(MsgType::HelloAck, seq, payload);
}

std::vector<uint8_t> encode_clock_sync_request(uint32_t seq, int64_t client_send_time_ns) {
	std::vector<uint8_t> payload;
	put_i64be(payload, client_send_time_ns);
	return wrap(MsgType::ClockSyncRequest, seq, payload);
}

std::vector<uint8_t> encode_clock_sync_response(uint32_t seq, int64_t client_send_time_ns, int64_t server_recv_time_ns, int64_t server_send_time_ns) {
	std::vector<uint8_t> payload;
	put_i64be(payload, client_send_time_ns);
	put_i64be(payload, server_recv_time_ns);
	put_i64be(payload, server_send_time_ns);
	return wrap(MsgType::ClockSyncResponse, seq, payload);
}

std::vector<uint8_t> encode_video_config(uint32_t seq, const std::string &json_body, const std::vector<uint8_t> &sps_pps_annexb) {
	std::vector<uint8_t> payload;
	put_u16be(payload, static_cast<uint16_t>(json_body.size()));
	put_bytes(payload, json_body);
	put_bytes(payload, sps_pps_annexb.data(), sps_pps_annexb.size());
	return wrap(MsgType::VideoConfig, seq, payload);
}

std::vector<uint8_t> encode_video_chunk(uint32_t seq, int64_t capture_timestamp_ns, bool is_keyframe, const uint8_t *nal_data, size_t nal_size) {
	std::vector<uint8_t> payload;
	put_i64be(payload, capture_timestamp_ns);
	put_u8(payload, is_keyframe ? 0x01 : 0x00);
	put_bytes(payload, nal_data, nal_size);
	return wrap(MsgType::VideoChunk, seq, payload);
}

std::vector<uint8_t> encode_imu_batch(uint32_t seq, const std::vector<ImuSample> &samples) {
	std::vector<uint8_t> payload;
	put_u16be(payload, static_cast<uint16_t>(samples.size()));
	for (const ImuSample &s : samples) {
		put_u8(payload, s.sensor_type);
		put_i64be(payload, s.timestamp_ns);
		put_f32be(payload, s.x);
		put_f32be(payload, s.y);
		put_f32be(payload, s.z);
	}
	return wrap(MsgType::ImuBatch, seq, payload);
}

std::vector<uint8_t> encode_pose_sample(uint32_t seq, int64_t timestamp_ns, uint8_t tracking_state, float x, float y, float z, float qx, float qy, float qz, float qw) {
	std::vector<uint8_t> payload;
	put_i64be(payload, timestamp_ns);
	put_u8(payload, tracking_state);
	put_f32be(payload, x);
	put_f32be(payload, y);
	put_f32be(payload, z);
	put_f32be(payload, qx);
	put_f32be(payload, qy);
	put_f32be(payload, qz);
	put_f32be(payload, qw);
	return wrap(MsgType::PoseSample, seq, payload);
}

std::vector<uint8_t> encode_point_cloud(uint32_t seq, int64_t timestamp_ns, const std::vector<Point> &points) {
	std::vector<uint8_t> payload;
	put_i64be(payload, timestamp_ns);
	put_u32be(payload, static_cast<uint32_t>(points.size()));
	for (const Point &pt : points) {
		put_f32be(payload, pt.x);
		put_f32be(payload, pt.y);
		put_f32be(payload, pt.z);
		put_f32be(payload, pt.confidence);
	}
	return wrap(MsgType::PointCloud, seq, payload);
}

std::vector<uint8_t> encode_camera_intrinsics(uint32_t seq, float fx, float fy, float cx, float cy, uint32_t width, uint32_t height) {
	std::vector<uint8_t> payload;
	put_f32be(payload, fx);
	put_f32be(payload, fy);
	put_f32be(payload, cx);
	put_f32be(payload, cy);
	put_u32be(payload, width);
	put_u32be(payload, height);
	return wrap(MsgType::CameraIntrinsics, seq, payload);
}

std::vector<uint8_t> encode_status(uint32_t seq, const std::string &json_body) {
	std::vector<uint8_t> payload;
	put_bytes(payload, json_body);
	return wrap(MsgType::Status, seq, payload);
}

std::vector<uint8_t> encode_goodbye(uint32_t seq, const std::string &json_body) {
	std::vector<uint8_t> payload;
	put_bytes(payload, json_body);
	return wrap(MsgType::Goodbye, seq, payload);
}

bool decode_clock_sync_request(const uint8_t *payload, size_t size, int64_t &out_client_send_time_ns) {
	return get_i64be(payload, size, 0, out_client_send_time_ns);
}

bool decode_clock_sync_response(const uint8_t *payload, size_t size, int64_t &out_client_send_time_ns, int64_t &out_server_recv_time_ns, int64_t &out_server_send_time_ns) {
	return get_i64be(payload, size, 0, out_client_send_time_ns) &&
			get_i64be(payload, size, 8, out_server_recv_time_ns) &&
			get_i64be(payload, size, 16, out_server_send_time_ns);
}

bool decode_imu_batch(const uint8_t *payload, size_t size, std::vector<ImuSample> &out_samples) {
	uint16_t count;
	if (!get_u16be(payload, size, 0, count)) {
		return false;
	}
	size_t offset = 2;
	out_samples.clear();
	out_samples.reserve(count);
	for (uint16_t i = 0; i < count; i++) {
		if (offset + 21 > size) {
			return false;
		}
		ImuSample s;
		s.sensor_type = payload[offset];
		if (!get_i64be(payload, size, offset + 1, s.timestamp_ns)) {
			return false;
		}
		if (!get_f32be(payload, size, offset + 9, s.x)) {
			return false;
		}
		if (!get_f32be(payload, size, offset + 13, s.y)) {
			return false;
		}
		if (!get_f32be(payload, size, offset + 17, s.z)) {
			return false;
		}
		out_samples.push_back(s);
		offset += 21;
	}
	return true;
}

bool decode_pose_sample(const uint8_t *payload, size_t size, int64_t &out_timestamp_ns, uint8_t &out_tracking_state, float &x, float &y, float &z, float &qx, float &qy, float &qz, float &qw) {
	if (size < 37) {
		return false;
	}
	if (!get_i64be(payload, size, 0, out_timestamp_ns)) {
		return false;
	}
	out_tracking_state = payload[8];
	return get_f32be(payload, size, 9, x) &&
			get_f32be(payload, size, 13, y) &&
			get_f32be(payload, size, 17, z) &&
			get_f32be(payload, size, 21, qx) &&
			get_f32be(payload, size, 25, qy) &&
			get_f32be(payload, size, 29, qz) &&
			get_f32be(payload, size, 33, qw);
}

bool decode_point_cloud(const uint8_t *payload, size_t size, int64_t &out_timestamp_ns, std::vector<Point> &out_points) {
	uint32_t count;
	if (!get_i64be(payload, size, 0, out_timestamp_ns)) {
		return false;
	}
	if (!get_u32be(payload, size, 8, count)) {
		return false;
	}
	size_t offset = 12;
	out_points.clear();
	out_points.reserve(count);
	for (uint32_t i = 0; i < count; i++) {
		if (offset + 16 > size) {
			return false;
		}
		Point p;
		if (!get_f32be(payload, size, offset, p.x) ||
				!get_f32be(payload, size, offset + 4, p.y) ||
				!get_f32be(payload, size, offset + 8, p.z) ||
				!get_f32be(payload, size, offset + 12, p.confidence)) {
			return false;
		}
		out_points.push_back(p);
		offset += 16;
	}
	return true;
}

bool decode_camera_intrinsics(const uint8_t *payload, size_t size, float &fx, float &fy, float &cx, float &cy, uint32_t &width, uint32_t &height) {
	if (size < 24) {
		return false;
	}
	return get_f32be(payload, size, 0, fx) &&
			get_f32be(payload, size, 4, fy) &&
			get_f32be(payload, size, 8, cx) &&
			get_f32be(payload, size, 12, cy) &&
			get_u32be(payload, size, 16, width) &&
			get_u32be(payload, size, 20, height);
}

} // namespace arstream::protocol
